File: algorithms/sorting/merge-sort/merge-sort.mjs
function merge(left, right) {
  const output = [];
  let leftIndex = 0;
  let rightIndex = 0;

  while (leftIndex < left.length && rightIndex < right.length) {
    if (left[leftIndex] <= right[rightIndex]) output.push(left[leftIndex++]);
    else output.push(right[rightIndex++]);
  }

  return output.concat(left.slice(leftIndex), right.slice(rightIndex));
}

/** Return a stable sorted copy of the input using Merge Sort. */
export function mergeSort(numbers) {
  if (numbers.length <= 1) return [...numbers];
  const middle = Math.floor(numbers.length / 2);
  const left = mergeSort(numbers.slice(0, middle));
  const right = mergeSort(numbers.slice(middle));
  return merge(left, right);
}

File: algorithms/sorting/merge-sort/merge_sort.py
def merge(left, right):
    output = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            output.append(left[left_index])
            left_index += 1
        else:
            output.append(right[right_index])
            right_index += 1

    return output + left[left_index:] + right[right_index:]


def merge_sort(numbers):
    """Return a stable sorted copy of numbers using Merge Sort."""
    if len(numbers) <= 1:
        return numbers.copy()

    middle = len(numbers) // 2
    left = merge_sort(numbers[:middle])
    right = merge_sort(numbers[middle:])
    return merge(left, right)
